src/platforms/stlinkv3/usb_f723.c

```diff
@@ -0,0 +1,178 @@
+/* ??? The st header files have this:
+ * #define USB_HS_PHYC_LDO_ENABLE                   USB_HS_PHYC_LDO_DISABLE
+ * ...go figure...
+ */
+#define OTG_PHYC_LDO_DISABLE (1 << 2)
+#define OTG_PHYC_LDO_STATUS  (1 << 1)
+
+/* Yes, this is unpleasant. It does not belong here. */
+#define _REG_BIT(base, bit) (((base) << 5) + (bit))
+
+/* STM32F7x3xx and STM32F730xx devices have an internal usb high-speed usb phy controller */
+enum {
+	RCC_OTGPHYC = _REG_BIT(0x44, 31),
+};
+
+/* Receive FIFO size in 32-bit words. */
+#define RX_FIFO_SIZE 512
+
+static usbd_device *stm32f723_usbd_init(void);
+
+static struct _usbd_device usbd_dev;
+
+const struct _usbd_driver stm32f723_usb_driver = {
+	.init = stm32f723_usbd_init,
+	.set_address = dwc_set_address,
+	.ep_setup = dwc_ep_setup,
+	.ep_reset = dwc_endpoints_reset,
+	.ep_stall_set = dwc_ep_stall_set,
+	.ep_stall_get = dwc_ep_stall_get,
+	.ep_nak_set = dwc_ep_nak_set,
+	.ep_write_packet = dwc_ep_write_packet,
+	.ep_read_packet = dwc_ep_read_packet,
+	.poll = dwc_poll,
+	.disconnect = dwc_disconnect,
+	.base_address = USB_OTG_HS_BASE,
+	.set_address_before_status = 1,
+	.rx_fifo_size = RX_FIFO_SIZE,
+};
+
+/*
+ * Initialize the USB device controller hardware of the STM32.
+ *
+ * Note: this initialization code was compiled from the libopencm3 usb_f207.c
+ * source, and the st cube sources. Note that the st manuals state that some delays
+ * are necessary at certain places. This code works for usb hosts that I tested on
+ * without the delays, but this is bending the rules.
+ *
+ * If for someone the code below does not work, one thing to try is perhaps to
+ * enable the delays before starting to debug other stuff. */
+static usbd_device *stm32f723_usbd_init(void)
+{
+	rcc_periph_clock_enable(RCC_GPIOB);
+	gpio_mode_setup(GPIOB, GPIO_MODE_AF, GPIO_PUPD_NONE, GPIO14 | GPIO15);
+	gpio_set_output_options(GPIOB, GPIO_OTYPE_PP, GPIO_OSPEED_100MHZ, GPIO14 | GPIO15);
+	gpio_set_af(GPIOB, GPIO_AF12, GPIO14 | GPIO15);
+
+	rcc_periph_clock_enable((enum rcc_periph_clken)RCC_OTGPHYC);
+	rcc_periph_clock_enable(RCC_OTGHSULPI);
+
+	// TODO - check the preemption and subpriority, they are unified here
+	nvic_set_priority(NVIC_OTG_HS_IRQ, 0);
+
+	rcc_periph_clock_enable(RCC_OTGHS);
+	OTG_HS_GINTSTS = OTG_GINTSTS_MMIS;
+
+	// ??? What is this??? It is not documented in the manual
+	OTG_HS_GCCFG |= OTG_GCCFG_PHYHSEN;
+
+	/* ??? The st header files have this:
+	 * #define USB_HS_PHYC_LDO_ENABLE                   USB_HS_PHYC_LDO_DISABLE
+	 * ...go figure...
+	 */
+	OTG_HS_PHYC_LDO |= OTG_PHYC_LDO_DISABLE;
+	while (!(OTG_HS_PHYC_LDO & OTG_PHYC_LDO_STATUS))
+		;
+	/* This setting is for a HSE clock of 25 MHz. */
+	OTG_HS_PHYC_PLL1 = 5 << 1;
+	OTG_HS_PHYC_TUNE |= 0x00000f13U;
+	OTG_HS_PHYC_PLL1 |= OTG_PHYC_PLL1_ENABLE;
+	/* 2ms Delay required to get internal phy clock stable
+	 * Used by DFU too, so platform_xxx not available.
+	 * Some Stlinkv3-Set did not cold start w/o the delay
+	 */
+	volatile int i = 200 * 1000;
```

Comment:
Let's turn this into a for loop with a `volatile size_t` loop counter counting up to the value assigned here